### PR TITLE
PSBT proprietary key system matching BIP 174

### PR DESCRIPTION
**Pull request by dr-orlovsky**
This is a companion PR to https://github.com/rust-bitcoin/rust-bitcoin/pull/478 and a part of epic https://github.com/rust-bitcoin/rust-bitcoin/issues/473

This PR is based on previous PRs for BIP 32 binary serialization and versioning from #469  and #470 (which must be merged first) and includes commits from them. Commits specific to this PR starts from https://github.com/rust-bitcoin/rust-bitcoin/pull/471/commits/325f323aebc201903d0839bb969f3cb2dcd325e2

Once #478 will be merged, I will add constant key name for the new hash key types into this PR

### CODE REVIEW

**Comment by stevenroose** [comment]
Again, could this perhaps be rebased on master?

**Comment by dr-orlovsky** [comment]
@stevenroose done

**Comment by dr-orlovsky** [comment]
Build fails b/c something strange with fuzztests, can't understand what's wrong

File: src/util/psbt/map/input.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ pub struct Input {
     pub witness_script: Option<Script>,
     /// A map from public keys needed to sign this input to their corresponding
     /// master key fingerprints and derivation paths.
-    pub hd_keypaths: BTreeMap<PublicKey, KeySource>,
+    pub bip32_derivation: BTreeMap<PublicKey, KeySource>,
```
Comment (stevenroose) [inline src/util/psbt/map/input.rs:74]: Hmm, is this a necessary breaking change? I recognize that there's a bunch of PSBT related breaking changes being combined, just wondering if this one is significant enough.

I do like the mention of bip32 in there, though, so ack.

File: src/util/psbt/serialize.rs [hunk 2]
```diff
@@ -70,7 +71,20 @@ impl Deserialize for PublicKey {
...
+}
+
+impl Deserialize for ExtendedPubKey {
+    fn deserialize(bytes: &[u8]) -> Result<Self, Error> {
+        ExtendedPubKey::decode(&bytes)
+            .map_err(|_| Error::ParseFailed("xpub parse error"))
+    }
+}
+
+impl Serialize for (Fingerprint, DerivationPath) {
```
Comment (stevenroose) [inline src/util/psbt/serialize.rs:87]: Why was this changed back from using `KeySource` to the composite type?

File: src/util/psbt/raw.rs [hunk 3]
```diff
@@ -119,3 +136,62 @@ impl Decodable for Pair {
...
+        len += e.write(&self.key)?;
+        Ok(len)
+    }
+}
+
+impl<Subtype> Decodable for ProprietaryKey<Subtype> where Subtype: Copy + From<u8> + Into<u8> {
+    fn consensus_decode<D: io::Read>(mut d: D) -> Result<Self, encode::Error> {
+        let VarInt(prefix_size) = VarInt::consensus_decode(&mut d)?;
+
+        let mut prefix = vec![0u8; prefix_size as usize];
```
Comment (stevenroose) [inline src/util/psbt/raw.rs:165]: There is a `Decodable` implementation for `Vec<u8>`.. So you can just do `let prefix = Vec::<u8>::consensus_decode(&mut d)?;`.

File: src/util/psbt/raw.rs [hunk 4]
```diff
@@ -119,3 +136,62 @@ impl Decodable for Pair {
         })
     }
 }
+
+impl From<u8> for ProprietaryType {
+    fn from(_: u8) -> Self {
+        ProprietaryType
```
Comment (stevenroose) [inline src/util/psbt/raw.rs:142]: This will not roundtrip. I think if you want a default type, it should be `pub struct ProprietaryType(u8);` so that you can roundtrip the actual byte.

File: src/util/psbt/raw.rs [hunk 5]
```diff
@@ -41,6 +41,23 @@ pub struct Pair {
...
 }
 
+/// Default implementation for proprietary key subtyping
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub struct ProprietaryType;
+
+/// Proprietary keys (i.e. keys starting with 0xFC byte) with their internal
+/// structure according to BIP 174.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub struct ProprietaryKey<Subtype = ProprietaryType> where Subtype: Copy + From<u8> + Into<u8> {
```
Comment (stevenroose) [inline src/util/psbt/raw.rs:51]: Alternatively, you could make `ProprietaryKeyType` a marker trait:
```rust
pub trait ProprietaryTypeKey: Copy + From<u8> + Into<u8> {}
```
But the plethora of naming will soon become confusing (`ProprietaryType`, `ProprietaryTypeKey`, `ProprietaryKey`).

**Review by stevenroose — CHANGES_REQUESTED** [review]
I think the fuzz failure is due to the bug that I pointed out in the `ProprietaryKey` type.

File: src/util/psbt/map/input.rs (hunk 1 above)
Comment (dr-orlovsky) [reply src/util/psbt/map/input.rs:74]: This is a part of the other PR which is needed to be accepted first https://github.com/rust-bitcoin/rust-bitcoin/pull/513

The rationale is to get all names matching BIP spec

File: src/util/psbt/serialize.rs (hunk 2 above)
Comment (dr-orlovsky) [reply src/util/psbt/serialize.rs:87]: This is a mistake, will change. The code was just based on an older version before that PR with `KeySource` was merged

File: src/util/psbt/raw.rs (hunk 5 above)
Comment (dr-orlovsky) [reply src/util/psbt/raw.rs:53]: If we would not use generic here this will allow using of conflicting proprietary key type systems, which is undesirable

File: src/util/psbt/raw.rs (hunk 4 above)
Comment (dr-orlovsky) [reply src/util/psbt/raw.rs]: Thanks, fixed

File: src/util/psbt/mod.rs [hunk 6]
```diff
@@ -612,6 +615,7 @@ mod tests {
                 },
                 version: 0,
                 xpub: Default::default(),
+                proprietary: Default::default(),
```
Comment (stevenroose) [inline src/util/psbt/mod.rs:618]: nit: You have mixed usage of `Default::default()` and `BTreeMap::new()` in this file.

**Review by stevenroose — APPROVED** [review]
LGTM

**Review by apoelstra — APPROVED** [review]
ack 7400bccb60a7d07c04e53d614814096e7ed203d6
